### ISSUE TITLE
Average training loss increases drastically every 10,000 iterations

### ISSUE THREAD

**Issue by kmatzen** [BVLC/caffe#1445]
Has anyone seen this?  Training loss is pretty low.  Testing loss is much higher which might indicate poor generalization.  But then every 10,000 iterations, the training loss jumps up, ending up near the testing loss.  I have the validation set running every 1,000 iterations and checkpoints every 10,000, but I can't reproduce the problem by making those much smaller.

BTW, for these logs, I increased the precision with which the loss is printed.  This is https://github.com/BVLC/caffe/commit/eca019aca7efce5293bb3a821b4b20cb401e384e after addressing https://github.com/BVLC/caffe/issues/1430 (although it happens with or without this fix).

```
129982  43763.226886  1.3238074779510498   0.001
129983  43764.307195  1.47053658962249756  0.001
129984  43765.425303  1.7798844575881958   0.001
129985  43766.505411  1.38557446002960205  0.001
129986  43767.584995  1.12471497058868408  0.001
129987  43768.664405  1.38750505447387695  0.001
129988  43769.744163  1.25835120677947998  0.001
129989  43770.824172  1.51574957370758057  0.001
129990  43771.904043  1.43048596382141113  0.001
129991  43772.983652  1.64656448364257812  0.001
129992  43774.063747  1.50741243362426758  0.001
129993  43775.143725  1.1661369800567627   0.001
129994  43776.223486  1.34615623950958252  0.001
129995  43777.303355  1.44158244132995605  0.001
129996  43778.383100  1.33253633975982666  0.001
129997  43779.462991  1.73792111873626709  0.001
129998  43780.543159  1.25823259353637695  0.001
129999  43781.623327  1.1047205924987793   0.001
130000  43799.186681  2.65336370468139648  0.001
130001  43800.265882  2.68952560424804688  0.001
130002  43801.345167  2.69243741035461426  0.001
130003  43802.424489  2.45040369033813477  0.001
130004  43803.503798  2.36295747756958008  0.001
130005  43804.583015  2.55356001853942871  0.001
130006  43805.662784  2.3663630485534668   0.001
130007  43806.742365  2.33002066612243652  0.001
130008  43807.821769  2.39991593360900879  0.001
130009  43808.900976  3.47261166572570801  0.001
130010  43809.980541  2.84621167182922363  0.001
130011  43811.059983  2.38281464576721191  0.001
130012  43812.139537  3.0418097972869873   0.001
130013  43813.219772  2.62189078330993652  0.001
130014  43814.299240  2.57108259201049805  0.001
130015  43815.378793  2.79705286026000977  0.001
```

```
I1117 14:29:07.386688 10300 solver.cpp:213] Iteration 128998, loss = 1.77805042266845703
I1117 14:29:07.386701 10300 solver.cpp:228]     Train net output #0: loss = 1.77805 (* 1 = 1.77805 loss)
I1117 14:29:07.386710 10300 solver.cpp:449] Iteration 128998, lr = 0.001
I1117 14:29:08.466507 10300 solver.cpp:213] Iteration 128999, loss = 2.14356589317321777
I1117 14:29:08.466521 10300 solver.cpp:228]     Train net output #0: loss = 2.14357 (* 1 = 2.14357 loss)
I1117 14:29:08.466529 10300 solver.cpp:449] Iteration 128999, lr = 0.001
I1117 14:29:08.474407 10300 solver.cpp:268] Iteration 129000, Testing net (#0)
I1117 14:29:22.188505 10300 solver.cpp:319]     Test net output #0: accuracy = 0.427
I1117 14:29:22.188616 10300 solver.cpp:319]     Test net output #1: loss = 2.67234 (* 1 = 2.67234 loss)
I1117 14:29:23.081106 10300 solver.cpp:213] Iteration 129000, loss = 1.61280250549316406
I1117 14:29:23.081120 10300 solver.cpp:228]     Train net output #0: loss = 1.6128 (* 1 = 1.6128 loss)
I1117 14:29:23.081128 10300 solver.cpp:449] Iteration 129000, lr = 0.001
I1117 14:29:24.160897 10300 solver.cpp:213] Iteration 129001, loss = 1.60916078090667725
I1117 14:29:24.160909 10300 solver.cpp:228]     Train net output #0: loss = 1.60916 (* 1 = 1.60916 loss)
I1117 14:29:24.160918 10300 solver.cpp:449] Iteration 129001, lr = 0.001
```

```
I1117 14:47:20.258363 10300 solver.cpp:213] Iteration 129997, loss = 1.73792111873626709
I1117 14:47:20.258375 10300 solver.cpp:228]     Train net output #0: loss = 1.73792 (* 1 = 1.73792 loss)
I1117 14:47:20.258384 10300 solver.cpp:449] Iteration 129997, lr = 0.001
I1117 14:47:21.338531 10300 solver.cpp:213] Iteration 129998, loss = 1.25823259353637695
I1117 14:47:21.338546 10300 solver.cpp:228]     Train net output #0: loss = 1.25823 (* 1 = 1.25823 loss)
I1117 14:47:21.338556 10300 solver.cpp:449] Iteration 129998, lr = 0.001
I1117 14:47:22.418699 10300 solver.cpp:213] Iteration 129999, loss = 1.1047205924987793
I1117 14:47:22.418712 10300 solver.cpp:228]     Train net output #0: loss = 1.10472 (* 1 = 1.10472 loss)
I1117 14:47:22.418720 10300 solver.cpp:449] Iteration 129999, lr = 0.001
I1117 14:47:23.462487 10300 solver.cpp:338] Snapshotting to <removed>
I1117 14:47:24.833917 10300 solver.cpp:346] Snapshotting solver state to <removed>
I1117 14:47:25.562151 10300 solver.cpp:268] Iteration 130000, Testing net (#0)
I1117 14:47:39.089681 10300 solver.cpp:319]     Test net output #0: accuracy = 0.43
I1117 14:47:39.089839 10300 solver.cpp:319]     Test net output #1: loss = 2.55749 (* 1 = 2.55749 loss)
I1117 14:47:39.982053 10300 solver.cpp:213] Iteration 130000, loss = 2.65336370468139648
I1117 14:47:39.982067 10300 solver.cpp:228]     Train net output #0: loss = 2.65336 (* 1 = 2.65336 loss)
I1117 14:47:39.982077 10300 solver.cpp:449] Iteration 130000, lr = 0.001
I1117 14:47:41.061254 10300 solver.cpp:213] Iteration 130001, loss = 2.68952560424804688
I1117 14:47:41.061266 10300 solver.cpp:228]     Train net output #0: loss = 2.68953 (* 1 = 2.68953 loss)
I1117 14:47:41.061275 10300 solver.cpp:449] Iteration 130001, lr = 0.001
I1117 14:47:42.140539 10300 solver.cpp:213] Iteration 130002, loss = 2.69243741035461426
I1117 14:47:42.140552 10300 solver.cpp:228]     Train net output #0: loss = 2.69244 (* 1 = 2.69244 loss)
I1117 14:47:42.140561 10300 solver.cpp:449] Iteration 130002, lr = 0.001
```

![loss-90000-140000](https://cloud.githubusercontent.com/assets/90536/5073891/8f93a8fa-6e53-11e4-8242-ae0997927bf8.png)

**Comment by gcr** [comment]
This could be caused by forgetting to pass `-shuffle` to the `create_imageset` tool, assuming that your dataset is exactly 10,000 batches large.

This could also be an artifact of the learning rate step size, Most example models I've seen set that value to 10,000 by default.

I'm curious to see if you found the solution yet :)

**Comment by gcr** [comment]
For the record, i've found that leaving out `-shuffle` tends to have super weird effects like this, especially if your training set is grouped by eg. label. Are the first `N` batches all a certain label that's hard to classify correctly? If so, that's likely related to what's going on here.

(My opinion: `-shuffle` should be the default. It's bit my research group more than a few times)

**Comment by shelhamer** [comment]
Closing as a one-off mystery. Feel free to discuss further on the mailing list.
